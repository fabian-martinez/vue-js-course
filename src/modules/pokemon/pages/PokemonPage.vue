<template>
    <div>
        <h1>Pokemon Page <span>No. {{ pokemonid }}</span></h1>
        <div v-if="pokemon">
            <img :src="pokemon.sprites.front_default" :alt="pokemon.name" srcset="">
        </div>
    </div>
</template>

<script>

    export default {
        props: {
            pokemonid:{
                type: Number,
                required: true
            },
        }, 
        data() {
            return{
                pokemon: null,
            }
        },
        created() { 
            //const { pokemonid } = this.$route.params;
            this.getPokemon()
        },
        methods: {
            async getPokemon() {
                try {
                    const pokemon = await fetch(`https://pokeapi.co/api/v2/pokemon/${ this.pokemonid }`  ).then( response => response.json() )
                    console.log(pokemon);
                    this.pokemon = pokemon 
                } catch (error) {
                     this.$router.push('/')
                }
            }
        },
        watch: {
             pokemonid() {
                this.getPokemon()
             }
        }
    }
</script>