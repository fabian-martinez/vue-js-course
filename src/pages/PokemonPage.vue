<template>
    <h1 v-if="!pokemon">Espere por favor!!!</h1>
    <div v-else>
      <h1>Â¿Quien es este pokemon?</h1>
      <pokemon-picture 
        :pokemonId="pokemon.id" 
        :showPokemon="showResult"/>
      <pokemon-options 
        :pokemons="pokemonArr"
        @pokemonSelected="getAnswer"/>
      <template v-if="showResult">
        <h1 class="fade-in">{{result}}</h1>
        <button @click="newGame">Iniciar de nuevo</button>
      </template>
    </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonPotions'

export default {
  components: { PokemonPicture, PokemonOptions },
  data() {
    return{
      pokemonArr: [],
      pokemon: null,
      showResult: false,
      result:''
    }
  },
  methods:{
    async mixPokemonArr() {
      this.pokemonArr = await getPokemonOptions()
      const rndInt = Math.floor( Math.random() * 4)
      this.pokemon = this.pokemonArr[rndInt]
    },
    getAnswer(idSelected){
      this.showResult = true
      if (idSelected === this.pokemon.id) {
        this.result = `Correcto, ${this.pokemon.name}`
      }else{
        this.result = `Error, era ${this.pokemon.name}`
      }
    },
    newGame(){
      this.pokemonArr = []
      this.showResult = false
      this.pokemon = null
      this.mixPokemonArr()

    }
  },
  mounted(){
    this.mixPokemonArr()
  }
}
</script>
