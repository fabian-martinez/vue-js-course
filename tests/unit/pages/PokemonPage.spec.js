import { shallowMount, mount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import { pokemonsArray } from '../mocks/pokemons.mock';


describe('PokemonPage Pages', () => {
    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    test('debe llamar a mixPokemonArray al montar', () => {
        const mixPokemonArr = jest.spyOn(PokemonPage.methods, 'mixPokemonArr')
        shallowMount(PokemonPage)
        expect(mixPokemonArr).toHaveBeenCalledTimes(1)
    });

    test('debe de hacer match con el snapshop cuando cargen los pokemons', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonsArray,
                    pokemon: pokemonsArray[0],
                    showResult: false,
                    result: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    });
});