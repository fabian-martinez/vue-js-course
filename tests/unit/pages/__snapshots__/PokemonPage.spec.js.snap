// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Pages debe de hacer match con el snapshop cuando cargen los pokemons 1`] = `
<div>
  <h1>Â¿Quien es este pokemon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
